### PR TITLE
JSON schema contribution guide

### PR DESCRIPTION
**Pull request by aarnq**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [x] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change 
- [ ] kind/dev-change   
- [ ] kind/security     
- [ ] [kind/adr]()      

### What does this PR do / why do we need this PR?

This adds guidelines on how to write schema.

Closes https://github.com/elastisys/ck8s-issue-tracker/issues/282

#### Checklist

- [ ] Proper commit message prefix on all commits
  
- Change checks:
  - [ ] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
  - [ ] The change upgrades CRDs
  - [ ] The change updates the config *and* the schema
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [ ] The bug fix is covered by regression tests

### CODE REVIEW

File: config/schemas/README.md [hunk 1]
```diff
@@ -109,3 +109,98 @@ This can be enabled this in other repositories by editing the file `.vscode/sett
...
+
+#### Defaults and examples
+
+- Must be provided for all simple types, all `array` types, all compositional, conditional, and reusable `object` types.
+- Must be valid input to the schema, with the exception of `set-me`.
+- Should degrade gracefully, e.g. disabling optional components rather than enabling.
+
+### Structural
+
+These guidelines aims to help to create schemas that are easier to read and write.
```
Comment (Zash) [inline config/schemas/README.md:170]:
```suggestion
These guidelines aims to help to create schemas that are easier to read and write.

- Each value should have a single `type`
- Use `additionalProperties: false` to reject unspecified properties and typos.
```

File: config/schemas/README.md [hunk 2]
```diff
@@ -109,3 +109,98 @@ This can be enabled this in other repositories by editing the file `.vscode/sett
...
+
+- Must include a comment under the key `$comment` in the schema with the stanza "Schema imported from \[\<upstream-name\>\]\(\<upstream-link\>\).".
+- Must include a reference to upstream documentation.
+
+#### Compositions and Conditions
+
+Conditions here refers to the use of `allOf`, `anyOf`, and `oneOf` to define schema compositions and conditions.
+
+- Must set defaults as they can else not be fully resolved.
+- Should set examples as they can else not be fully resolved.
```
[the remark this answers is not in the document]
Comment (aarnq) [reply config/schemas/README.md:205]: Yes, I've reworded the list items.

File: config/schemas/README.md (hunk 1 above)
Comment (aarnq) [reply config/schemas/README.md:170]: Added.

File: config/schemas/README.md [hunk 3]
```diff
@@ -109,3 +109,155 @@ This can be enabled this in other repositories by editing the file `.vscode/sett
...
+- Must be provided, else it becomes rendered as `Untitled`.
+- Should be capitalised, as it becomes part of rendered headers `<title> Schema`.
+- Should be short, as it becomes part of rendered names.
+- Should be based on the key of the schema, as it becomes more intuitive.
+- Should include the context of the schema for common keys, e.g. `Object Storage` and `Object Storage Network Policies`.
+- Should include the context type of the schema last, as it becomes part of rendered headers `<title> <type> Schema`.
+
+#### Descriptions
+
+- Must be [GitHub Flavoured Markdown (GFM)](https://github.github.com/gfm/), as it will be rendered down into markdown documentation.
```
Comment (Xartos) [inline config/schemas/README.md:138]: Is this true? Isn't this rendered into mkdocs markdown? I think there's a slight difference, or at least has been

Comment (aarnq) [reply config/schemas/README.md:138]: Yes, but we run enough plugins to not make this matter, so it is easier to just use the same variant across this repository.

**Review by Xartos — APPROVED** [review]

**Pull request merged**